modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessContextAction.java

```diff
@@ -0,0 +1,327 @@
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.painless.lookup.PainlessClass;
+import org.elasticsearch.painless.lookup.PainlessClassBinding;
+import org.elasticsearch.painless.lookup.PainlessConstructor;
+import org.elasticsearch.painless.lookup.PainlessField;
+import org.elasticsearch.painless.lookup.PainlessInstanceBinding;
+import org.elasticsearch.painless.lookup.PainlessLookup;
+import org.elasticsearch.painless.lookup.PainlessLookupUtility;
+import org.elasticsearch.painless.lookup.PainlessMethod;
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.RestToXContentListener;
+import org.elasticsearch.script.ScriptContext;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+
+/**
+ * Internal REST API for querying context information about Painless whitelists.
+ * Commands include the following:
+ * <ul>
+ *     <li> GET /_scripts/painless/_context -- retrieves a list of contexts </li>
+ *     <li> GET /_scripts/painless/_context?context=%name% --
+ *     retrieves all available information about the API for this specific context</li>
+ * </ul>
+ */
+public class PainlessContextAction extends Action<PainlessContextAction.Response> {
+
+    static final PainlessContextAction INSTANCE = new PainlessContextAction();
+    private static final String NAME = "cluster:admin/scripts/painless/context";
+
+    private static final String SCRIPT_CONTEXT_NAME_PARAM = "context";
+
+    private PainlessContextAction() {
+        super(NAME);
+    }
+
+    @Override
+    public Response newResponse() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Writeable.Reader<Response> getResponseReader() {
+        return Response::new;
+    }
+
+    public static class Request extends ActionRequest {
+
+        private String scriptContextName;
+
+        public Request() {
+            scriptContextName = null;
+        }
+
+        public void setScriptContextName(String scriptContextName) {
+            this.scriptContextName = scriptContextName;
+        }
+
+        public String getScriptContextName() {
+            return scriptContextName;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
+
+        private final PainlessScriptEngine painlessScriptEngine;
```

Comment:
I think that `PainlessScriptEngine` should't be exposed in this response class. This something internal to painless itself. I think it makes more sense to introduce a simple pojo classes that are exposed here and that can serialize the information that users of this api need. The transport action can then built the response class with the pojo classes based on the painless script engine instance.

Reply:
I considered this and wasn't sure it was the right thing to do since so much information is being copied from the PainlessLookup.  But based on your feedback it makes sense for the serialization and encapsulation, so I'll work on changing it.  Thanks for this feedback!

Reply:
Added several POJOs to accommodate this.